### ISSUE TITLE
Breaking changes in v0.9.3

### ISSUE THREAD

**Issue by sethaxen** [JuliaDiff/FiniteDifferences.jl#67]
v0.9.3 introduced some breaking changes

Setup:
```julia
julia> using LinearAlgebra, FiniteDifferences, StaticArrays
julia> x = Float64[1, 2, 3]
julia> fdm = central_fdm(5, 1)
julia> x2 = SVector{3}(x);
```

### Jacobian of the identity function no longer produces the identity matrix.
```julia
julia> jacobian(fdm, identity, x)[1] # expected identity matrix
```

on v0.1.2:
```julia
3×3 Array{Float64,2}:
 1.0          6.50515e-15  6.50515e-15
 1.30103e-14  1.0          1.30103e-14
 0.0          0.0          1.0
```

on v0.1.3:
```julia
3×3 Array{Float64,2}:
 4.16334e-16  4.16334e-16  4.16334e-16
 8.32667e-16  8.32667e-16  8.32667e-16
 0.0          0.0          0.0
```

### StaticArrays no longer supported

```julia
julia> jacobian(fdm, identity, x2)[1]
```

on v0.1.2:
```julia
 3×3 Array{Float64,2}:
  1.0          6.50515e-15  6.50515e-15
  1.30103e-14  1.0          1.30103e-14
  0.0          0.0          1.0
```

on v0.1.3:
```julia
ERROR: MethodError: no method matching to_vec(::SArray{Tuple{3},Float64,1,3})
Closest candidates are:
  to_vec(::Number) at /Users/saxen/.julia/packages/FiniteDifferences/kqb1h/src/to_vec.jl:8
  to_vec(::Array{#s121,1} where #s121<:Number) at /Users/saxen/.julia/packages/FiniteDifferences/kqb1h/src/to_vec.jl:15
  to_vec(::Array{T,1} where T) at /Users/saxen/.julia/packages/FiniteDifferences/kqb1h/src/to_vec.jl:17
  ...
Stacktrace:
 [1] jacobian(::FiniteDifferences.Central{UnitRange{Int64},Array{Float64,1}}, ::Function, ::SArray{Tuple{3},Float64,1,3}; len::Nothing) at /Users/saxen/.julia/packages/FiniteDifferences/kqb1h/src/grad.jl:28
 [2] jacobian(::FiniteDifferences.Central{UnitRange{Int64},Array{Float64,1}}, ::Function, ::SArray{Tuple{3},Float64,1,3}) at /Users/saxen/.julia/packages/FiniteDifferences/kqb1h/src/grad.jl:28
 [3] top-level scope at REPL[34]:1
```

**Comment by willtebbutt** [comment]
Hmmm yeah, these are definitely bugs. I'll take a look at sorting them out. Thanks for raising the issue!

**Comment by willtebbutt** [comment]
So now that I've dug around a little, it turns out that the `identity` issue is definitely a bug, but the `StaticArrays` thing actually isn't.

In 0.9.3 we tightened up the `to_vec` abstraction a little. It used to define a `to_vec` method for `AbstractArray`s. However, this causes problems for defining the `from_vec` methods when there are fewer degrees of freedom in your array than there are elements. Consider, for example, `Fills` from `FillArrays`. They can be of an arbitrary size, but have only a single d.o.f. If you try and use the same `from_vec` implementation for a `Fill` as a `Vector`, you're bound to find up with problems. Before, we had silent failures and difficult to understand bugs.

Now we have much more restrictive definitions of `to_vec`, that will have to be manually extended to handle new subtypes of `AbstractArray`. I'll take a look at doing this for `StaticArrays` via `Requires` tomorrow.

It's pretty annoying that `to_vec` doesn't happen for free any more for array types that the general implementation makes sense for, but I think it's probably necessary.

**Comment by sethaxen** [comment]
Since the API documents that `jacobian` can be used with `AbstractArray{<:Real}` and did before this release, by semvar shouldn't it be 0.10.0? See https://www.juliadiff.org/FiniteDifferences.jl/latest/pages/api/#FiniteDifferences.jacobian. This was a breaking change for our package, and we were using FiniteDifferences in the documented way.

**Comment by nickrobinson251** [comment]
I agree that dropping support for StaticArrays (presuming they worked fine before) is a breaking change. We should revert that and tag v0.9.4. Then reapply the change (with the identity fix) as v0.10

**Comment by willtebbutt** [comment]
Agreed -- this was a mistake on my part. Thanks again for bringing this up @sethaxen

**Comment by nickrobinson251** [comment]
https://github.com/JuliaDiff/FiniteDifferences.jl/pull/68 now fixes the StaticArray thing too, so we can just tag a new patch, right?

That's great work -- thanks, @willtebbutt!

**Comment by willtebbutt** [comment]
Resolved by #68 , which has now been merged and 0.9.4 is tagged.
